libdnf/rpm/solv/id_queue.hpp

```diff
@@ -0,0 +1,119 @@
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_RPM_SOLV_ID_QUEUE_HPP
+#define LIBDNF_RPM_SOLV_ID_QUEUE_HPP
+
+#include <utility>
+
+extern "C" {
+#include <solv/queue.h>
+}
+
+namespace libdnf::rpm::solv {
+
+struct IdQueue {
+public:
+    IdQueue();
+    IdQueue(const IdQueue & src);
+    IdQueue(IdQueue && src);
+    explicit IdQueue(const Queue & src);
+    ~IdQueue();
+
+    bool operator==(const IdQueue & other) const;
+    bool operator!=(const IdQueue & other) const;
+    IdQueue & operator=(const IdQueue & src);
+    IdQueue & operator=(IdQueue && src) noexcept;
+    Id operator[](int index) const;
+
+    void push_back(Id id);
+    void push_back(Id id1, Id id2);
+    int * data() const noexcept;
+    Queue & get_queue() noexcept;
+    int size() const noexcept;
+    void clear() noexcept;
+    void insert(const IdQueue & src);
+    void reserve(int n);
+
+private:
+    Queue queue;
+
+};
+
+inline IdQueue::IdQueue() { queue_init(&queue); }
+inline IdQueue::IdQueue(const IdQueue & src) { queue_init_clone(&queue, &src.queue); }
+inline IdQueue::IdQueue(IdQueue && src)
+{
+    queue_init(&queue);
+    std::swap(queue, src.queue);
+}
+inline IdQueue::IdQueue(const Queue & src) { queue_init_clone(&queue, &src); }
+
+inline IdQueue::~IdQueue() { queue_free(&queue); }
+
+inline bool IdQueue::operator==(const IdQueue & other) const
+{
+    if (size() != other.size()) {
+        return false;
+    }
+
+    for (int i = 0; i < size(); i++) {
+        if ((*this)[i] != other[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+inline bool IdQueue::operator!=(const IdQueue & other) const { return !(*this == other); }
+
+inline IdQueue & IdQueue::operator=(const IdQueue & src)
+{
+    if (this == &src) {
+        return *this;
+    }
+    queue_empty(&queue);
+    insert(src);
+    return *this;
+}
+inline IdQueue & IdQueue::operator=(IdQueue && src) noexcept
+{
+    std::swap(queue, src.queue);
+    return *this;
+}
+
+inline void IdQueue::push_back(Id id) { queue_push(&queue, id); }
+inline void IdQueue::push_back(Id id1, Id id2) { queue_push2(&queue, id1, id2); }
+inline Id IdQueue::operator[](int index) const { return queue.elements[index]; }
+inline int * IdQueue::data() const noexcept { return queue.elements; }
+inline Queue & IdQueue::get_queue() noexcept { return queue; }
+inline int IdQueue::size() const noexcept { return queue.count; }
+inline void IdQueue::clear() noexcept { queue_empty(&queue); }
+
+inline void
+IdQueue::insert(const IdQueue & src)
```

Comment:
What does the insert do? I'd expect it inserting an Id at a specified position.

Reply:
It inserts the new data at the end position. I can rename it, or add an argument, or overloaded, or add description. What do you think?

Reply:
Based on the argument being another an IdQueue containing multiple values, I expect it extends the original queue.
I'd use `extend` in that case.

Reply:
Renamed to append()